### PR TITLE
 Add C implementation for `@stdlib/number/float64/base/signbit`

### PR DESCRIPTION
**Pull request by Pranavchiku**
## Checklist

- [x] include/stdlib/number/float64/base
- [x] include.gypi
- [x] manifest.json
- [x] binding.gyp
- [x] src
- [x] examples
- [x] lib
- [x] benchmark
- [x] test
- [x] update readme.md
* * *

@kgryte

### CODE REVIEW

**Comment by Pranavchiku** [comment]
@kgryte this PR can have a final review :)

File: lib/node_modules/@stdlib/number/float64/base/signbit/README.md [hunk 1]
```diff
@@ -78,6 +78,95 @@ for ( i = 0; i < 100; i++ ) {
...
+stdlib_base_float64_signbit( 3.14, &out );
+```
+
+The function accepts the following arguments:
+
+-   **x**: `[in] double` input value.
+-   **out**: `[out] int8_t*` destination for unbiased signbit.
+
+```c
+void stdlib_base_float64_signbit( const double x, int8_t *out );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/number/float64/base/signbit/README.md]:
```suggestion
int8_t stdlib_base_float64_signbit( const double x );
```

File: lib/node_modules/@stdlib/number/float64/base/signbit/README.md [hunk 2]
```diff
@@ -78,6 +78,95 @@ for ( i = 0; i < 100; i++ ) {
...
+#include <stdint.h>
+
+int8_t out;
+stdlib_base_float64_signbit( 3.14, &out );
+```
+
+The function accepts the following arguments:
+
+-   **x**: `[in] double` input value.
+-   **out**: `[out] int8_t*` destination for unbiased signbit.
```
Comment (kgryte) [inline lib/node_modules/@stdlib/number/float64/base/signbit/README.md]:
```suggestion
```

File: lib/node_modules/@stdlib/number/float64/base/signbit/README.md [hunk 3]
```diff
@@ -78,6 +78,95 @@ for ( i = 0; i < 100; i++ ) {
...
+
+### Usage
+
+```c
+#include "stdlib/number/float64/base/signbit.h"
+```
+
+#### stdlib_base_float64_signbit( x, \*out )
+
+Extracts a boolean indicating if the sign bit for a double-precision floating-point number is on (true) or off (false).
```
Comment (kgryte) [inline lib/node_modules/@stdlib/number/float64/base/signbit/README.md]:
```suggestion
Returns an integer indicating whether the sign bit for a double-precision floating-point number is on (`1`) or off (`0`).
```

File: lib/node_modules/@stdlib/number/float64/base/signbit/README.md [hunk 4]
```diff
@@ -78,6 +78,95 @@ for ( i = 0; i < 100; i++ ) {
...
+#include <stdio.h>
+
+int main() {
+    double x[] = { 3.14, -3.14, 0.0, -0.0, 4.0, 1.0, -1.0, 1.0e308, -1.0e308 };
+
+    int8_t out;
+    int i;
+    for ( i = 0; i < 9; i++ ) {
+        stdlib_base_float64_signbit( x[ i ], &out );
+        printf( "%lf => signbit: %u\n", x[ i ], out );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/number/float64/base/signbit/README.md]: Nit: need to use macro for printing `int8_t` values.

File: lib/node_modules/@stdlib/number/float64/base/signbit/README.md [hunk 5]
```diff
@@ -78,6 +78,95 @@ for ( i = 0; i < 100; i++ ) {
...
+}
+```
+
+</section>
+
+<!-- /.examples -->
+
+</section>
+
+<!-- /.examples -->
```
Comment (kgryte) [inline lib/node_modules/@stdlib/number/float64/base/signbit/README.md]:
```suggestion
<!-- /.c -->
```

File: lib/node_modules/@stdlib/number/float64/base/signbit/benchmark/benchmark.native.js [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+var signbit = tryRequire( resolve( __dirname, './../lib/native.js' ) );
+var opts = {
+	'skip': ( signbit instanceof Error )
+};
+
+
+// MAIN //
+
+bench( pkg, opts, function benchmark( b ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/number/float64/base/signbit/benchmark/benchmark.native.js]:
```suggestion
bench( pkg+'::native', opts, function benchmark( b ) {
```

File: lib/node_modules/@stdlib/number/float64/base/signbit/benchmark/benchmark.native.js [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+	var x;
+	var y;
+	var i;
+
+	b.tic();
+	for ( i = 0; i < b.iterations; i++ ) {
+		x = ( randu()*1.0e7 ) - 5.0e6;
+		y = signbit( x );
+		if ( y !== 1 || y !== 0 ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/number/float64/base/signbit/benchmark/benchmark.native.js]:
```suggestion
		if ( y !== 1 && y !== 0 ) {
```

File: lib/node_modules/@stdlib/number/float64/base/signbit/benchmark/benchmark.native.js [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+			b.fail( 'should return 0 or 1' );
+		}
+	}
+	b.toc();
+	if ( y !== 1 || y !== 0 ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/number/float64/base/signbit/benchmark/benchmark.native.js]:
```suggestion
	if ( y !== 1 && y !== 0 ) {
```

File: lib/node_modules/@stdlib/number/float64/base/signbit/benchmark/c/benchmark.c [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+	int8_t out;
+	double x;
+	double t;
+	int i;
+
+	t = tic();
+	for ( i = 0; i < ITERATIONS; i++ ) {
+		x = ( rand_double()*200.0 ) - 100.0;
+		stdlib_base_float64_signbit( x, &out );
+		if ( out != (int8_t)signbit( x ) ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/number/float64/base/signbit/benchmark/c/benchmark.c]:
```suggestion
		if ( out != 0 && out != 1 ) {
```

File: lib/node_modules/@stdlib/number/float64/base/signbit/examples/c/example.c [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+#include <stdio.h>
+
+int main() {
+	double x[] = { 3.14, -3.14, 0.0, -0.0, 4.0, 1.0, -1.0, 1.0e308, -1.0e308 };
+
+	int8_t out;
+	int i;
+	for ( i = 0; i < 9; i++ ) {
+		stdlib_base_float64_signbit( x[ i ], &out );
+		printf( "%lf => signbit: %u\n", x[ i ], out );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/number/float64/base/signbit/examples/c/example.c]: Nit: use macro for `int8_t`

File: lib/node_modules/@stdlib/number/float64/base/signbit/include/stdlib/number/float64/base/signbit.h [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+/*
+* If C++, prevent name mangling so that the compiler emits a binary file having undecorated names, thus mirroring the behavior of a C compiler.
+*/
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+
+/**
+* Extracts a boolean indicating if the sign bit for a double-precision floating-point number is on (true) or off (false).
```
Comment (kgryte) [inline lib/node_modules/@stdlib/number/float64/base/signbit/include/stdlib/number/float64/base/signbit.h]: Update description and signature.

File: lib/node_modules/@stdlib/number/float64/base/signbit/lib/native.js [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+
+var addon = require( './../lib/addon.node' );
+
+
+// MAIN //
+
+/**
+* Returns a boolean indicating if the sign bit is on (true) or off (false).
+*
+* @param {number} x - input value
```
Comment (kgryte) [inline lib/node_modules/@stdlib/number/float64/base/signbit/lib/native.js:33]:
```suggestion
* @private
* @param {number} x - input value
```

File: lib/node_modules/@stdlib/number/float64/base/signbit/lib/native.js [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+* @returns {integer8} - boolean indicating if sign bit is on or off
```
Comment (kgryte) [inline lib/node_modules/@stdlib/number/float64/base/signbit/lib/native.js]:
```suggestion
* @returns {boolean} - boolean indicating if sign bit is on or off
```

File: lib/node_modules/@stdlib/number/float64/base/signbit/lib/native.js [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+* @example
+* var bool = signbit( 0.0 );
+* // returns false
+*
+* @example
+* var bool = signbit( -0.0 );
+* // returns true
+*/
+function signbit( x ) {
+	return addon( x );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/number/float64/base/signbit/lib/native.js]:
```suggestion
	return Boolean( addon( x ) );
```

File: lib/node_modules/@stdlib/number/float64/base/signbit/lib/native.js [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+'use strict';
+
+// MODULES //
+
+var addon = require( './../lib/addon.node' );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/number/float64/base/signbit/lib/native.js]:
```suggestion
var Boolean = require( '@stdlib/boolean/ctor' );
var addon = require( './../src/addon.node' );
```

File: lib/node_modules/@stdlib/number/float64/base/signbit/src/signbit.c [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+#include "stdlib/number/float64/base/signbit.h"
+#include "stdlib/number/float64/base/get_high_word.h"
+#include <stdint.h>
+
+/**
+* Extracts a boolean indicating if the sign bit for a double-precision floating-point number is on (true) or off (false).
```
Comment (kgryte) [inline lib/node_modules/@stdlib/number/float64/base/signbit/src/signbit.c]: Update desc and signature.

File: lib/node_modules/@stdlib/number/float64/base/signbit/src/signbit.c [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+*
+* @param x       input value
+* @param out     destination for the extracted boolean
```
Comment (kgryte) [inline lib/node_modules/@stdlib/number/float64/base/signbit/src/signbit.c]:
```suggestion
* @param x       input value
* @return          result
```

File: lib/node_modules/@stdlib/number/float64/base/signbit/src/signbit.c [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+*
+* @example
+* #include <stdint.h>
+*
+* int8_t out;
+* stdlib_base_float64_signbit( 3.14, &out );
+*/
+void stdlib_base_float64_signbit( const double x, int8_t *out ) {
+	uint32_t high;
+	// Extract from the input value a higher order word (unsigned 32-bit integer) containing the exponent and sign:
```
Comment (kgryte) [inline lib/node_modules/@stdlib/number/float64/base/signbit/src/signbit.c:38]:
```suggestion

	// Extract from the input value a higher order word (unsigned 32-bit integer) containing the exponent and sign:
```

File: lib/node_modules/@stdlib/number/float64/base/signbit/test/test.native.js [hunk 19]
```diff
@@ -0,0 +1,77 @@
...
+var tape = require( 'tape' );
+var randu = require( '@stdlib/random/base/randu' );
+var round = require( '@stdlib/math/base/special/round' );
+var pow = require( '@stdlib/math/base/special/pow' );
+var tryRequire = require( '@stdlib/utils/try-require' );
+
+
+// VARIABLES //
+
+var signbit = tryRequire( resolve( __dirname, './../lib/native.js' ) );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/number/float64/base/signbit/test/test.native.js:33]: To compile native add-ons, use the following command:

```bash
make install-node-addons NODE_ADDONS_PATTERN=float64/base/signbit
```

**Review by kgryte — CHANGES_REQUESTED** [review]
Thanks for working on this, @Pranavchiku! Just some minor nits. Once resolved, this should be ready for merge!

**Comment by Pranavchiku** [comment]
I guess all the changes are done, and when I try to run the `test.native.js` locally, it throws me the following log
```
(base) pranavchiku@pranavchiku-Yoga-Slim-7-14IIL05:~/stdlib(signbit1)$ node /home/pranavchiku/stdlib/lib/node_modules/@stdlib/number/float64/base/signbit/test/test.native.js
TAP version 13
# main export is a function
ok 1 /home/pranavchiku/stdlib/lib/node_modules/@stdlib/number/float64/base/signbit/test/test.native.js
ok 2 main export is a function
# the function returns a boolean
node: symbol lookup error: /home/pranavchiku/stdlib/lib/node_modules/@stdlib/number/float64/base/signbit/src/addon.node: undefined symbol: napi_create_int8
```

File: lib/node_modules/@stdlib/number/float64/base/signbit/src/addon.c [hunk 20]
```diff
@@ -0,0 +1,92 @@
...
+
+	double value;
+	status = napi_get_value_double( env, argv[ 0 ], &value );
+	assert( status == napi_ok );
+
+	int8_t out;
+	out = stdlib_base_float64_signbit( value );
+
+	napi_value v;
+	status = napi_create_int8( env, out, &v );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/number/float64/base/signbit/src/addon.c]:
```suggestion
	status = napi_create_int32( env, out, &v );
```

N-API does not have an `int8` API. `int32` should be fine to use here.

Comment (Pranavchiku) [reply lib/node_modules/@stdlib/number/float64/base/signbit/src/addon.c]: ohh okay, got it :)

Comment (kgryte) [reply lib/node_modules/@stdlib/number/float64/base/signbit/src/addon.c]: For reference: https://nodejs.org/api/n-api.html#functions-to-convert-from-c-types-to-node-api

File: lib/node_modules/@stdlib/number/float64/base/signbit/lib/native.js [hunk 21]
```diff
@@ -0,0 +1,60 @@
...
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+'use strict';
+
+// MODULES //
+
+var Boolean = require( '@stdlib/boolean/ctor' );
+var Boolean = require( '@stdlib/boolean/ctor' );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/number/float64/base/signbit/lib/native.js]:
```suggestion
var Boolean = require( '@stdlib/boolean/ctor' );
```

**Comment by Pranavchiku** [comment]
Okay so, I am currently having my previously generated `addon.node`, and it has`napi_create_int8`, when I run that file again it throws me error, regarding `#include "stdlib/number/float64/base/signbit.h"`, how may I update that `addon.node`?

**Comment by kgryte** [comment]
Just rerun the command to build the addon. It will remove the old one and rebuild.

**Comment by Pranavchiku** [comment]
>Just rerun the command to build the addon. It will remove the old one and rebuild.

It generated during the time I pushed the files to PR, else to run locally I used `gcc <path>` which throws me error.

**Comment by kgryte** [comment]
For running locally, you should also use "make install-node-addons". This is necessary in order to resolve headers and deps.

**Comment by Pranavchiku** [comment]
Wait let me check, I already did that, trying it out again.

**Comment by kgryte** [comment]
In the project, we don't compile with gcc directly. We use GYP, which is what Node.js uses as its meta build tool.

**Comment by Pranavchiku** [comment]
Yeah, all Test cases passed, thanks @kgryte!

**Comment by Pranavchiku** [comment]
This PR can have a final review :)

File: lib/node_modules/@stdlib/number/float64/base/signbit/README.md [hunk 22]
```diff
@@ -78,6 +78,94 @@ for ( i = 0; i < 100; i++ ) {
...
+
+<section class="usage">
+
+### Usage
+
+```c
+#include "stdlib/number/float64/base/signbit.h"
+```
+
+#### stdlib_base_float64_signbit( x, \*out )
```
Comment (kgryte) [inline lib/node_modules/@stdlib/number/float64/base/signbit/README.md]:
```suggestion
#### stdlib_base_float64_signbit( x )
```

File: lib/node_modules/@stdlib/number/float64/base/signbit/README.md [hunk 23]
```diff
@@ -78,6 +78,94 @@ for ( i = 0; i < 100; i++ ) {
...
+
+Returns an integer indicating whether the sign bit for a double-precision floating-point number is on (`1`) or off (`0`).
+
+```c
+#include <stdint.h>
+
+int8_t out;
+stdlib_base_float64_signbit( 3.14, &out );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/number/float64/base/signbit/README.md]:
```suggestion
stdlib_base_float64_signbit( 3.14 );
```

File: lib/node_modules/@stdlib/number/float64/base/signbit/README.md [hunk 24]
```diff
@@ -78,6 +78,94 @@ for ( i = 0; i < 100; i++ ) {
...
+int main() {
+    double x[] = { 3.14, -3.14, 0.0, -0.0, 4.0, 1.0, -1.0, 1.0e308, -1.0e308 };
+
+    int8_t out;
+    int i;
+    for ( i = 0; i < 9; i++ ) {
+        stdlib_base_float64_signbit( x[ i ], &out );
+        printf( "%lf => signbit: %u\n", x[ i ], out );
+    }
+}
```
Comment (kgryte) [inline lib/node_modules/@stdlib/number/float64/base/signbit/README.md]:
```suggestion
#include "stdlib/number/float64/base/signbit.h"
#include <stdint.h>
#include <stdio.h>
#include <inttypes.h>

int main() {
    double x[] = { 3.14, -3.14, 0.0, -0.0, 4.0, 1.0, -1.0, 1.0e308, -1.0e308 };

    int8_t out;
    int i;
    for ( i = 0; i < 9; i++ ) {
        stdlib_base_float64_signbit( x[ i ], &out );
        printf( "%lf => signbit: %" PRId8 "\n", x[ i ], out );
    }
}
```

File: lib/node_modules/@stdlib/number/float64/base/signbit/benchmark/c/benchmark.c [hunk 25]
```diff
@@ -0,0 +1,137 @@
...
+	for ( i = 0; i < ITERATIONS; i++ ) {
+		x = ( rand_double()*200.0 ) - 100.0;
+		out = stdlib_base_float64_signbit( x );
+		if ( out != 0 && out != 1 ) {
+			printf( "unexpected result\n" );
+			break;
+		}
+	}
+	elapsed = tic() - t;
+	if ( out != (int8_t)signbit( x ) ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/number/float64/base/signbit/benchmark/c/benchmark.c]:
```suggestion
	if ( out != 0 && out !== 1 ) {
```

C99 [`signbit`](https://en.cppreference.com/w/c/numeric/math/signbit) is not guaranteed to return `1` for negative values.

File: lib/node_modules/@stdlib/number/float64/base/signbit/examples/c/example.c [hunk 26]
```diff
@@ -0,0 +1,33 @@
...
+#include <stdio.h>
+
+int main() {
+	double x[] = { 3.14, -3.14, 0.0, -0.0, 4.0, 1.0, -1.0, 1.0e308, -1.0e308 };
+
+	int8_t out;
+	int i;
+	for ( i = 0; i < 9; i++ ) {
+		out = stdlib_base_float64_signbit( x[ i ] );
+		printf( "%lf => signbit: %u\n", x[ i ], out );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/number/float64/base/signbit/examples/c/example.c]:
```suggestion
#include "stdlib/number/float64/base/signbit.h"
#include <stdint.h>
#include <stdio.h>
#include <inttypes.h>

int main() {
	double x[] = { 3.14, -3.14, 0.0, -0.0, 4.0, 1.0, -1.0, 1.0e308, -1.0e308 };

	int8_t out;
	int i;
	for ( i = 0; i < 9; i++ ) {
		out = stdlib_base_float64_signbit( x[ i ] );
		printf( "%lf => signbit: %" PRId8 "\n", x[ i ], out );
```

File: lib/node_modules/@stdlib/number/float64/base/signbit/src/addon.c [hunk 27]
```diff
@@ -0,0 +1,92 @@
...
+		assert( status == napi_ok );
+		return NULL;
+	}
+
+	double value;
+	status = napi_get_value_double( env, argv[ 0 ], &value );
+	assert( status == napi_ok );
+
+	int8_t out;
+	out = stdlib_base_float64_signbit( value );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/number/float64/base/signbit/src/addon.c]:
```suggestion
	int8_t out = stdlib_base_float64_signbit( value );
```

File: lib/node_modules/@stdlib/number/float64/base/signbit/src/addon.c [hunk 28]
```diff
@@ -0,0 +1,92 @@
...
+
+	napi_value v;
+	status = napi_create_int32( env, out, &v );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/number/float64/base/signbit/src/addon.c]:
```suggestion
	status = napi_create_int32( env, (int32_t)out, &v );
```

File: lib/node_modules/@stdlib/number/float64/base/signbit/src/signbit.c [hunk 29]
```diff
@@ -0,0 +1,44 @@
...
+*
+* @private
+* @param x       input value
+* @return        result
+*
+* @example
+* #include <stdint.h>
+*
+* int8_t out;
+* stdlib_base_float64_signbit( 3.14, &out );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/number/float64/base/signbit/src/signbit.c]:
```suggestion
* int8_t out = stdlib_base_float64_signbit( 3.14 );
```

**Review by kgryte — CHANGES_REQUESTED** [review]

File: lib/node_modules/@stdlib/number/float64/base/signbit/README.md [hunk 30]
```diff
@@ -78,6 +78,95 @@ for ( i = 0; i < 100; i++ ) {
...
+
+#### stdlib_base_float64_signbit( x )
+
+Returns an integer indicating whether the sign bit for a double-precision floating-point number is on (`1`) or off (`0`).
+
+```c
+#include <stdint.h>
+
+int8_t out;
+stdlib_base_float64_signbit( 3.14 );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/number/float64/base/signbit/README.md]:
```suggestion
int8_t out = stdlib_base_float64_signbit( 3.14 );
```

**Review by kgryte — APPROVED** [review]
Thanks, @Pranavchiku!

**Pull request merged**